File: apps/frontend/src-tauri/src/commands/docker.rs
//! Docker image management commands
//!
//! Commands for checking image availability and pulling images.

use crate::models::AppError;
use crate::services::ApiClient;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Helper to get an authenticated API client
fn get_client() -> Result<ApiClient, AppError> {
    ApiClient::new()
}

// =============================================================================
// Types
// =============================================================================

/// Image availability status for a single flavor
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FlavorImageStatus {
    pub available: bool,
    pub image_name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub size: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Response from flavor images endpoint
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FlavorImagesResponse {
    pub success: bool,
    pub images: HashMap<String, FlavorImageStatus>,
}

/// Docker image info
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DockerImageInfo {
    pub name: String,
    pub id: String,
    pub size: u64,
    pub created: String,
}

/// Response for single image check
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ImageExistsResponse {
    pub exists: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub image_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub image: Option<DockerImageInfo>,
}

/// Response for image pull (sync)
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ImagePullResponse {
    pub success: bool,
    pub image_name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub image: Option<DockerImageInfo>,
}

/// Docker daemon info
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DockerImageInfo2 {
    pub version: String,
    pub api_version: String,
    pub os: String,
    pub arch: String,
    pub containers: DockerContainerStats,
    pub images: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DockerContainerStats {
    pub running: u32,
    pub stopped: u32,
}

/// Docker info response wrapper
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DockerInfoResponse {
    pub success: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub info: Option<DockerImageInfo2>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Docker health check response
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DockerHealthCheckResponse {
    pub healthy: bool,
    pub timestamp: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Image pull request body
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ImagePullRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub image_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub flavor_id: Option<String>,
}

// =============================================================================
// Commands
// =============================================================================

/// Get availability status of all flavor images
#[tauri::command]
pub async fn get_flavor_images() -> Result<FlavorImagesResponse, AppError> {
    let client = get_client()?;
    client.get("/api/docker/images/flavors").await
}

/// Check if a specific image exists
#[tauri::command]
pub async fn check_image_exists(image_name: String) -> Result<ImageExistsResponse, AppError> {
    let client = get_client()?;
    // URL-encode the image name to handle colons in tags
    let encoded_name = urlencoding::encode(&image_name);
    client.get(&format!("/api/docker/images/{}", encoded_name)).await
}

/// Pull an image synchronously (without streaming progress)
#[tauri::command]
pub async fn pull_image_sync(
    image_name: Option<String>,
    flavor_id: Option<String>,
) -> Result<ImagePullResponse, AppError> {
    if image_name.is_none() && flavor_id.is_none() {
        return Err(AppError::InvalidConfig("Either image_name or flavor_id is required".to_string()));
    }
    
    let client = get_client()?;
    
    let body = ImagePullRequest {
        image_name,
        flavor_id,
    };
    
    let response: ImagePullResponse = client.post("/api/docker/images/pull/sync", &body).await?;
    
    if !response.success {
        return Err(AppError::ApiError(response.error.unwrap_or_else(|| "Failed to pull image".to_string())));
    }
    
    Ok(response)
}

/// Get Docker daemon information
#[tauri::command]
pub async fn get_docker_info() -> Result<DockerImageInfo2, AppError> {
    let client = get_client()?;
    
    let response: DockerInfoResponse = client.get("/api/docker/info").await?;
    
    if !response.success {
        return Err(AppError::ApiError(response.error.unwrap_or_else(|| "Failed to get Docker info".to_string())));
    }
    
    response.info.ok_or_else(|| AppError::ApiError("No Docker info returned".to_string()))
}

/// Check Docker daemon health
#[tauri::command]
pub async fn check_docker_health() -> Result<DockerHealthCheckResponse, AppError> {
    let client = get_client()?;
    client.get("/api/docker/health").await
}

File: apps/hub/src/routes/docker.ts
/**
 * Docker Image Management API Routes
 * Endpoints for checking image availability and pulling images
 */

import { Hono } from "hono";
import { streamSSE } from "hono/streaming";
import { getSandboxManager } from "../services/sandbox-manager.ts";
import { createLogger } from "../utils/logger.ts";

const log = createLogger("docker-routes");

export const dockerRouter = new Hono();

// =============================================================================
// GET /docker/images/flavors
// =============================================================================
// Get availability status of all flavor images
dockerRouter.get("/images/flavors", async (c) => {
  log.info("Checking flavor image availability");

  try {
    const manager = getSandboxManager();
    const availability = await manager.checkFlavorImageAvailability();

    return c.json({
      success: true,
      images: availability,
    });
  } catch (error) {
    log.error("Failed to check flavor image availability", { error });
    return c.json(
      {
        success: false,
        error: "Failed to check image availability",
      },
      500
    );
  }
});

// =============================================================================
// GET /docker/images/:imageName
// =============================================================================
// Check if a specific image exists
dockerRouter.get("/images/:imageName{.+}", async (c) => {
  const imageName = c.req.param("imageName");
  log.info("Checking image existence", { imageName });

  try {
    const manager = getSandboxManager();
    const imageInfo = await manager.getImageInfo(imageName);

    if (imageInfo) {
      return c.json({
        exists: true,
        image: {
          name: imageInfo.name,
          id: imageInfo.id,
          size: imageInfo.size,
          created: imageInfo.created,
        },
      });
    } else {
      return c.json({
        exists: false,
        imageName,
      });
    }
  } catch (error) {
    log.error("Failed to check image", { imageName, error });
    return c.json(
      {
        exists: false,
        error: "Failed to check image",
      },
      500
    );
  }
});

// =============================================================================
// POST /docker/images/pull
// =============================================================================
// Pull an image with SSE progress streaming
dockerRouter.post("/images/pull", async (c) => {
  const body = await c.req.json<{ imageName?: string; flavorId?: string }>();
  const { imageName, flavorId } = body;

  if (!imageName && !flavorId) {
    return c.json(
      {
        success: false,
        error: "Either imageName or flavorId is required",
      },
      400
    );
  }

  const manager = getSandboxManager();
  const targetImage = imageName || manager.getFlavorImageName(flavorId!);

  log.info("Starting image pull", { targetImage, flavorId });

  // Use SSE for streaming progress
  return streamSSE(c, async (stream) => {
    try {
      await stream.writeSSE({
        event: "start",
        data: JSON.stringify({ imageName: targetImage, status: "starting" }),
      });

      await manager.pullImage(targetImage, async (progress) => {
        await stream.writeSSE({
          event: "progress",
          data: JSON.stringify({
            status: progress.status,
            progress: progress.progress,
            id: progress.id,
          }),
        });
      });

      await stream.writeSSE({
        event: "complete",
        data: JSON.stringify({
          success: true,
          imageName: targetImage,
        }),
      });

      log.info("Image pull completed", { targetImage });
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : "Unknown error";
      log.error("Image pull failed", { targetImage, error: errorMessage });

      await stream.writeSSE({
        event: "error",
        data: JSON.stringify({
          success: false,
          error: errorMessage,
          imageName: targetImage,
        }),
      });
    }
  });
});

// =============================================================================
// POST /docker/images/pull/sync
// =============================================================================
// Pull an image synchronously (without streaming) - simpler for some use cases
dockerRouter.post("/images/pull/sync", async (c) => {
  const body = await c.req.json<{ imageName?: string; flavorId?: string }>();
  const { imageName, flavorId } = body;

  if (!imageName && !flavorId) {
    return c.json(
      {
        success: false,
        error: "Either imageName or flavorId is required",
      },
      400
    );
  }

  const manager = getSandboxManager();
  const targetImage = imageName || manager.getFlavorImageName(flavorId!);

  log.info("Starting synchronous image pull", { targetImage, flavorId });

  try {
    await manager.pullImage(targetImage);

    // Get image info after pull
    const imageInfo = await manager.getImageInfo(targetImage);

    log.info("Synchronous image pull completed", { targetImage });

    return c.json({
      success: true,
      imageName: targetImage,
      image: imageInfo
        ? {
            id: imageInfo.id,
            size: imageInfo.size,
            created: imageInfo.created,
          }
        : null,
    });
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : "Unknown error";
    log.error("Synchronous image pull failed", { targetImage, error: errorMessage });

    return c.json(
      {
        success: false,
        error: errorMessage,
        imageName: targetImage,
      },
      500
    );
  }
});

// =============================================================================
// GET /docker/info
// =============================================================================
// Get Docker daemon information
dockerRouter.get("/info", async (c) => {
  log.info("Getting Docker info");

  try {
    const manager = getSandboxManager();
    const info = await manager.getDockerInfo();

    return c.json({
      success: true,
      info,
    });
  } catch (error) {
    log.error("Failed to get Docker info", { error });
    return c.json(
      {
        success: false,
        error: "Failed to get Docker info",
      },
      500
    );
  }
});

// =============================================================================
// GET /docker/health
// =============================================================================
// Check Docker daemon health
dockerRouter.get("/health", async (c) => {
  try {
    const manager = getSandboxManager();
    const healthy = await manager.healthCheck();

    return c.json({
      healthy,
      timestamp: new Date().toISOString(),
    });
  } catch (error) {
    return c.json({
      healthy: false,
      error: "Docker health check failed",
      timestamp: new Date().toISOString(),
    });
  }
});
